File: ruby/binary-search/binary_search.rb
class BinarySearch
  attr_reader :list, :middle

  def initialize(array, check = true)
    if check
      array.each_cons(2) do |a, b|
        raise ArgumentError if a > b
      end
    end

    @list = array
    @middle = @list.size / 2
  end

  def search_for(n)
    if @list.size == 1
      if @list.first == n
        return 0
      else
        raise RuntimeError
      end
    end

    case @list[@middle] <=> n
    when -1
      BinarySearch.new(@list[@middle..], false).search_for(n) + @middle
    when 0
      @middle
    when 1
      BinarySearch.new(@list[...@middle], false).search_for(n)
    end
  end
end

File: c/binary-search/src/binary_search.c
#include "binary_search.h"

int* binary_search(int value, int* arr, size_t length) {
    if (length == 1) {
        if (arr[0] == value)
            return arr;
        else
            return NULL;
    }

    if (length == 0)
        return NULL;

    size_t middle = length / 2;

    if (arr[middle] < value)
        return binary_search(value, &arr[middle], middle + 1);
    else if (arr[middle] > value)
        return binary_search(value, arr, middle);
    else
        return &arr[middle];
}
